File: src/test/java/com/shelterhub/controller/MedicalRecordControllerTest.java
package com.shelterhub.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.shelterhub.dto.request.MedicalRecordRequest;
import com.shelterhub.dto.response.MedicalRecordResponse;
import com.shelterhub.service.MedicalRecordService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static com.shelterhub.utils.MedicalRecordUtils.buildMedicalRecord;
import static com.shelterhub.utils.MedicalRecordUtils.buildMedicalRecordRequest;
import static org.hamcrest.Matchers.hasSize;
import static org.mockito.Mockito.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(MedicalRecordController.class)
public class MedicalRecordControllerTest {

    private final String PATH_URL = "/v1/medical-record";

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private MedicalRecordService medicalRecordService;

    @Test
    public void shouldReturnAllMedicalRecords() throws Exception {
        MedicalRecordResponse firstMedicalRecord = buildMedicalRecordRequest().toMedicalRecord().toResponse();
        MedicalRecordResponse secondMedicalRecord = buildMedicalRecordRequest().toMedicalRecord().toResponse();

        when(medicalRecordService.getAll()).thenReturn(List.of(firstMedicalRecord, secondMedicalRecord));

        mockMvc.perform(MockMvcRequestBuilders.get(PATH_URL))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(jsonPath("$", hasSize(2)))
                .andExpect(jsonPath("$[0].id").value(firstMedicalRecord.getId().toString()))
                .andExpect(jsonPath("$[0].animalId").value(firstMedicalRecord.getAnimalId().toString()))
                .andExpect(jsonPath("$[1].id").value(secondMedicalRecord.getId().toString()))
                .andExpect(jsonPath("$[1].animalId").value(secondMedicalRecord.getAnimalId().toString()));

        verify(medicalRecordService, times(1)).getAll();
        verifyNoMoreInteractions(medicalRecordService);
    }

    @Test
    public void shouldReturnNotFoundWhenAllMedicalRecordsDoesNotExists() throws Exception {
        when(medicalRecordService.getAll()).thenReturn(new ArrayList<>());

        mockMvc.perform(MockMvcRequestBuilders.get(PATH_URL))
                .andExpect(status().isNotFound());

        verify(medicalRecordService, times(1)).getAll();
        verifyNoMoreInteractions(medicalRecordService);
    }


    @Test
    public void shouldReturnMedicalRecordById() throws Exception {
        var medicalRecordRequest = buildMedicalRecordRequest();
        var medicalRecordId = medicalRecordRequest.getId();

        var medicalResponse = medicalRecordRequest
                .toMedicalRecord()
                .toResponse();

        when(medicalRecordService.getById(medicalRecordId))
                .thenReturn(Optional.of(medicalResponse));

        mockMvc.perform(MockMvcRequestBuilders.get(PATH_URL + "/{id}", medicalRecordId))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(jsonPath("$.id").value(medicalResponse.getId().toString()))
                .andExpect(jsonPath("$.animalId").value(medicalResponse.getAnimalId().toString()));


        verify(medicalRecordService, times(1)).getById(medicalRecordId);
        verifyNoMoreInteractions(medicalRecordService);
    }

    @Test
    public void shouldReturnNotFoundWhenMedicalRecordDoesNotExist() throws Exception {
        var uuid = UUID.randomUUID();
        when(medicalRecordService.getById(uuid))
                .thenReturn(Optional.empty());

        mockMvc.perform(MockMvcRequestBuilders.get(PATH_URL + "/{id}", uuid.toString()))
                .andExpect(status().isNotFound());

        verify(medicalRecordService, times(1)).getById(uuid);
        verifyNoMoreInteractions(medicalRecordService);
    }

    @Test
    public void shouldCreateMedicalRecordSuccessfully() throws Exception {
        MedicalRecordRequest medicalRecordRequest = buildMedicalRecordRequest();
        var medicalRecord = medicalRecordRequest.toMedicalRecord();
        var medicalRecordResponse = medicalRecord.toResponse();

        when(medicalRecordService.create(medicalRecordRequest)).thenReturn(medicalRecordResponse);

        mockMvc.perform(MockMvcRequestBuilders.post(PATH_URL)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(new ObjectMapper().writeValueAsString(medicalRecordRequest)))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(jsonPath("$.id").value(medicalRecordResponse.getId().toString()))
                .andExpect(jsonPath("$.animalId").value(medicalRecordResponse.getAnimalId().toString()));

        verify(medicalRecordService, times(1)).create(any());
        verifyNoMoreInteractions(medicalRecordService);
    }

    @Test
    public void shouldUpdateMedicalRecordSuccessfully() throws Exception {
        var medicalRecordRequest = buildMedicalRecordRequest();
        var medicalRecordId = medicalRecordRequest.getId();
        var medicalRecord = medicalRecordRequest.toMedicalRecord();
        var medicalRecordResponse = medicalRecord.toResponse();

        when(medicalRecordService.update(medicalRecordRequest, medicalRecordId))
                .thenReturn(medicalRecordResponse);

        mockMvc.perform(MockMvcRequestBuilders.put(PATH_URL + "/{id}", medicalRecordId)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(new ObjectMapper().writeValueAsString(medicalRecordRequest)))
                .andExpect(status().isOk())
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(jsonPath("$.id").value(medicalRecordId.toString()))
                .andExpect(jsonPath("$.animalId").value(medicalRecordResponse.getAnimalId().toString()));

        verify(medicalRecordService, times(1)).update(medicalRecordRequest, medicalRecordId);
        verifyNoMoreInteractions(medicalRecordService);
    }

    @Test
    public void shouldDeleteMedicalRecordSuccessfully() throws Exception {
        var medicalRecord = buildMedicalRecord();
        var medicalRecordId = buildMedicalRecord().getId();

        when(medicalRecordService.deleteById(medicalRecordId))
                .thenReturn(Optional.of(medicalRecord));

        mockMvc.perform(MockMvcRequestBuilders.delete(PATH_URL + "/{id}", medicalRecordId))
                .andExpect(status().isNoContent());

        verify(medicalRecordService, times(1)).deleteById(any());
        verifyNoMoreInteractions(medicalRecordService);
    }
}

File: src/test/kotlin/com/shelterhub/controller/MedicalRecordControllerTest.kt
package com.shelterhub.controller

import br.com.shelterhubmanagementapi.controller.MedicalRecordController
import br.com.shelterhubmanagementapi.domain.model.toResponse
import br.com.shelterhubmanagementapi.dto.request.toMedicalRecord
import br.com.shelterhubmanagementapi.dto.response.AnimalResponse
import br.com.shelterhubmanagementapi.dto.response.MedicalRecordResponse
import br.com.shelterhubmanagementapi.exception.GlobalExceptionHandler
import br.com.shelterhubmanagementapi.service.MedicalRecordService
import com.ninjasquad.springmockk.MockkBean
import com.shelterhub.utils.MedicalRecordUtils
import io.mockk.Called
import io.mockk.coEvery
import io.mockk.coVerify
import kotlinx.coroutines.CompletableDeferred
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest
import org.springframework.http.MediaType
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.reactive.server.WebTestClient
import org.springframework.test.web.reactive.server.expectBody

@WebFluxTest(controllers = [MedicalRecordController::class])
@ContextConfiguration(classes = [MedicalRecordController::class, MedicalRecordService::class, GlobalExceptionHandler::class])
class MedicalRecordControllerTest(
    @Autowired private val webTestClient: WebTestClient,
) {
    private val PATH_URL = "/v1/medical-record"

    @MockkBean
    private lateinit var medicalRecordService: MedicalRecordService

    @Test
    fun `should return all medical records`() {
        val firstMedicalRecord: MedicalRecordResponse =
            MedicalRecordUtils.buildMedicalRecordResponse()

        val secondMedicalRecord: MedicalRecordResponse =
            MedicalRecordUtils.buildMedicalRecordResponse()

        coEvery { medicalRecordService.getAll() } returns CompletableDeferred(listOf(firstMedicalRecord, secondMedicalRecord))

        webTestClient.get()
            .uri { it.path(PATH_URL).build() }
            .exchange()
            .expectStatus().isOk
            .expectHeader().contentType(MediaType.APPLICATION_JSON)
            .expectBody<List<MedicalRecordResponse>>()

        coVerify(exactly = 1) { medicalRecordService.getAll() }
        coVerify { medicalRecordService.getAll() wasNot Called }
    }

    @Test
    fun `should return not found when all medical records does not exist`() {
        coEvery { medicalRecordService.getAll() } returns CompletableDeferred(emptyList())

        webTestClient.get()
            .uri { it.path(PATH_URL).build() }
            .exchange()
            .expectStatus().isNotFound

        coVerify(exactly = 1) { medicalRecordService.getAll() }
        coVerify { medicalRecordService.getAll() wasNot Called }
    }

    @Test
    fun `should return medical record by id`() {
        val medicalRecordRequest = MedicalRecordUtils.buildMedicalRecordRequest()
        val medicalRecordId = medicalRecordRequest.id
        val medicalResponse = MedicalRecordUtils.buildMedicalRecordResponse()

        coEvery { medicalRecordService.getById(medicalRecordId) } returns CompletableDeferred(medicalResponse)

        webTestClient.get().uri { uriBuilder ->
            uriBuilder
                .path("$PATH_URL/{id}")
                .build(medicalRecordId)
        }.exchange()
            .expectStatus().isOk
            .expectBody<MedicalRecordResponse>()

        coVerify { medicalRecordService.getById(medicalRecordId) wasNot Called }
    }

    @Test
    fun `should return not found when medical record does not exist`() {
        coEvery { medicalRecordService.getAll() } returns CompletableDeferred(emptyList())

        webTestClient.get()
            .uri { it.path(PATH_URL).build() }
            .exchange()
            .expectStatus().isNotFound

        coVerify(exactly = 1) { medicalRecordService.getAll() }
    }

    @Test
    fun `should create medical record successfully`() {
        val medicalRecordRequest = MedicalRecordUtils.buildMedicalRecordRequest()
        val medicalRecord = medicalRecordRequest.toMedicalRecord()
        val medicalRecordResponse = medicalRecord.toResponse()

        coEvery { medicalRecordService.create(medicalRecordRequest) } returns CompletableDeferred(medicalRecordResponse)

        webTestClient.post()
            .uri { it.path(PATH_URL).build() }
            .bodyValue(medicalRecordRequest)
            .exchange()
            .expectStatus().isCreated
            .expectHeader().contentType(MediaType.APPLICATION_JSON)
            .expectBody<AnimalResponse>()

        coVerify(exactly = 1) { medicalRecordService }
    }

    @Test
    fun `should update medical record successfully`() {
        val medicalRecordRequest = MedicalRecordUtils.buildMedicalRecordRequest()
        val medicalRecordId = medicalRecordRequest.id
        val medicalRecord = medicalRecordRequest.toMedicalRecord()
        val medicalRecordResponse = medicalRecord.toResponse()

        coEvery { medicalRecordService.update(medicalRecordRequest, medicalRecordId) } returns CompletableDeferred(medicalRecordResponse)

        webTestClient.post()
            .uri { it.path(PATH_URL).build() }
            .bodyValue(medicalRecordResponse)
            .exchange()
            .expectStatus().isCreated
            .expectHeader()
            .contentType(MediaType.APPLICATION_JSON)
            .expectBody<AnimalResponse>()

        coVerify { medicalRecordService.update(medicalRecordRequest, medicalRecordId) }
    }

    @Test
    fun `should delete medical record successfully`() {
        val medicalRecordId = MedicalRecordUtils.buildMedicalRecord().id

        webTestClient.delete()
            .uri { uriBuilder ->
                uriBuilder
                    .path("$PATH_URL/{id}")
                    .build(medicalRecordId)
            }.exchange()
            .expectStatus().isNotFound

        coVerify(exactly = 1) { medicalRecordService.deleteById(medicalRecordId) }
    }
}
